parsers/CA_SK.py

```diff
@@ -1,142 +1,107 @@
-
-
-def validate_zone_key(zone_key: str) -> None:
-    if zone_key != "CA-SK":
-        raise ParserException(
-            "CA_SK.py",
-            f"CA_SK.py is not designed to parse zone_key: {zone_key}.",
-            zone_key,
-        )
+USER_AGENT = (
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
+    "(KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36"
+)
 
 
-def validate_no_datetime(target_datetime: datetime | None, zone_key) -> None:
+def _request(
+    session: Session,
+    target_datetime: datetime | None,
+    url: str,
+    zone_key: ZoneKey,
+) -> dict | str:
+    # The source does not offer historical data, so bail out if it's requested.
     if target_datetime:
+        raise ParserException("CA_SK.py", "Unable to fetch historical data", zone_key)
+    # The zone key must be "CA-SK"; bail out otherwise.
+    if zone_key != "CA-SK":
+        raise ParserException("CA_SK.py", f"Cannot parse zone '{zone_key}'", zone_key)
+    # Mimic a user browser in the headers or the API will respond with a 403.
+    response = session.get(url, headers={"user-agent": USER_AGENT})
+    if not response.ok:
         raise ParserException(
             "CA_SK.py",
-            "This parser is unable to fetch historical data.",
+            f"Request to {url} failed. Response Code: {response.status_code}\n"
+            f"Error:\n{response.text}",
             zone_key,
         )
+    return response.json()
 
 
 def fetch_production(
-    zone_key: str = "CA-SK",
-    session: Session | None = None,
+    zone_key: ZoneKey = ZoneKey("CA-SK"),
+    session: Session = Session(),
     target_datetime: datetime | None = None,
     logger: Logger = getLogger(__name__),
-):
-    """This parser function will currently return the daily average of the day in question as hourly data.
-    This is because the API only returns daily data but the backend expects hourly values.
-    This is in order to facilitate the estimation of the hourly values from the daily average.
-    """
-    # Validate that the zone key is equal to CA-SK.
-    validate_zone_key(zone_key)
-    # Validate that the target_datetime is None as this parser is unable to fetch historical data.
-    validate_no_datetime(target_datetime, zone_key)
-
-    session = session or Session()
-
-    # Set the headers to mimic a user browser as the API will return a 403 if not.
-    headers = {"user-agent": USER_AGENT}
-    response: Response = session.get(PRODUCTION_URL, headers=headers)
-
-    if not response.ok:
-        raise ParserException(
-            "CA_SK.py",
-            f"Failed to fetch production data. Response Code: {response.status_code}\nError:\n{response.text}",
-            zone_key,
-        )
-
-    raw_data = response.json()
+) -> list[dict[str, Any]]:
+    payload = _request(session, target_datetime, PRODUCTION_URL, zone_key)
```

Comment:
```suggestion
    payload = _request(session, target_datetime, PRODUCTION_URL, zone_key)
    if isinstance(payload, str):
        raise ParserException("CA_SK.py", "Expected dict but returned payload was str", zone_key)
```

Since str was added as a return type we need to check that this is a dict for type narrowing before we can access it without type errors.

Reply:
By that logic, wouldn't we also want to check for a `dict` in the other parser function? I don't think there's any real need for a runtime type check, because we already know that the consumption URL always responds with a `str` and the production URL always returns a `dict`. The type hint alludes to this fact, but doesn't make it explicit.